Server implementation

silver-volt4 opened the pull request:
<!--

* Before making a pull request, ensure the changes are for an approved issue.
* If your changes are not for an approved issue, your pull request can and will be rejected.
*
* CHECK https://github.com/PretendoNetwork/REPO_NAME/issues?q=is%3Aopen+is%3Aissue+label%3Aapproved
* FOR APPROVED ISSUES!

-->

Resolves [this forum thread](https://forum.pretendo.network/t/we-implemented-swapdoodle-and-would-like-to-contribute-it/16204/6)

### Changes:

Adds an implementation for Swapdoodle. This uses Jon's unfinished DataStore rework (https://github.com/PretendoNetwork/nex-protocols-common-go/pull/53).

The server currently doesn't work _out of the box_ due to the incomplete state of the DataStore rework. There are some quirks I discovered that I wanted to discuss here, instead of polluting the DataStore PR with possibly Swapdoodle-specific info. This would be:
- [object_enabled.go:27](https://github.com/PretendoNetwork/nex-protocols-common-go/blob/ae941d444e29ee190f23f096ccccec8e77b1728d/datastore/database/object_enabled.go#L27): Swapdoodle seems to be uploading doodles with the `needUploadCompletion` flag set to false, which in turn causes [uploadCompleted to be set to true](https://github.com/PretendoNetwork/nex-protocols-common-go/blob/ae941d444e29ee190f23f096ccccec8e77b1728d/datastore/database/insert_object_by_prepare_post_param.go#L194), and so CompletePostObjectV1 fails. Additionally, since ObjectEnabled also checks for `state == 0`, the file is actually never accepted because it's [always initialized to 0](https://github.com/PretendoNetwork/nex-protocols-common-go/blob/ae941d444e29ee190f23f096ccccec8e77b1728d/datastore/database/insert_object_by_prepare_post_param.go#L187) but that's likely to be fixed anyway
- ~~[This callback](https://github.com/SwapdoodleRevival/swapdoodle/blob/6acba3f8d8e1889e4fc8d45e780923e9907158d0/nex/register_protocols.go#L43) is actually never run, because CompletePostObjectV1 does not call it. This likely won't be needed in the future when notification support is added to the package, I'm currently creating notifications manually for the sake of maintaining functionality~~ Notifications are now supported, so this is no longer relevant

With the two quirks patched out locally via a Go workspace, I have confirmed the server to correctly relay notes.
<!--

* Describe your changes in as much detail as possible. Make sure to list your changes, as well as the rationale behind them.
* If applicable, include code snippets, images, videos, etc.
*
* If your changes require any database migrations, describe them in detail and leave any migration queries inside of a code
* block within a <details> tag.

-->

<hr>

- [X] I have read and agreed to the [Code of Conduct](https://github.com/PretendoNetwork/Pretendo/blob/master/.github/CODE_OF_CONDUCT.md).
- [X] I have read and complied with the [contributing guidelines](https://github.com/PretendoNetwork/Pretendo/blob/master/.github/CONTRIBUTING.md).
- [ ] What I'm implementing was an [approved issue](../issues?q=is%3Aopen+is%3Aissue+label%3Aapproved). (it theoretically wasn't?)
- [X] I have tested all of my changes.

silver-volt4 committed:
Merge content with https://github.com/SwapdoodleRevival/hpp-server

This adds virtually all files from the repository mentioned above. Since the previous content is more of a template/skeleton, I did not bother with properly merging things.

silver-volt4 committed:
Update dependency files

silver-volt4 committed:
feat: use unfinished DataStore rework by Jon

Silver-Volt4 commented:
how i feel about this pr (lol):
![obrazek](https://github.com/user-attachments/assets/4a346a25-d941-43ee-9749-fab830eb75e9)

Silver-Volt4 commented:
Interesting... Thanks for the info, I haven't really dived into NEX much outside of Swapdoodle. Now I understand the upload process a bit more clearly.

[Here](https://github.com/user-attachments/files/20476175/flows.txt) is a mitmproxy capture of my 3DS uploading a note that I have just taken. One thing to note right away is the `version` header on the HPP requests, it is set to `NEX_2_4_1_S25`. Unless NEX/HPP uses some exotic versioning, I'd say that doesn't look like NEX 3.10.1 at all 😀 So at least that can be explained.

FYI, Mitmproxy makes it a bit tedious to extract the raw NEX packet from the HTTP request. I use this bash script (courtesy of @centdemeern1) to extract the packet from a hex dump (which is easy to view in mitm): 
```
xxd -r - | python3 -c 'import sys;s = sys.stdin.read();i = s.find("\r\n"*2)+4;sys.stdout.write(s[i:s.find("----------", i)-2]);sys.exit(0);' | xxd -p
```

Though honestly, I'm still not convinced that checking for ObjectEnabled during CompletePostObject is correct - unless all files uploaded have the DATA_FLAG_NEED_REVIEW flag set? Or maybe I'm just overlooking something obvious

DaniElectra commented:
HPP connections are made by the BOSS sysmodule, which bundles NEX by itself, so the version between that the games have and the version used on the HPP connection will be different. As the console sends on the request, HPP on the 3DS uses NEX 2.4.1 `NEX_2_4_1_S25`, but Swapdoodle is bundled with NEX 3.10.1 within tha game.

The same also applies on the Wii U, where the NEX version that is bundled in BOSS is 3.5.2, while Smash 4 has NEX version 3.6, so the HPP requests that the game does through BOSS will use version 3.5.2

Silver-Volt4 commented:
There's one more thing to Swapdoodle that I probably should have mentioned it sooner. I apologise if it causes any trouble.

In my testing, I've found that Swapdoodle _refuses_ to upload notes if it can't download `dstsetting` from BOSS. Returning a 304 is enough to make it work, even on a completely empty save file (when no `If-Modified-Since` is sent), which is what I ultimately ended up locally patching into the BOSS server to make Swapdoodle work when hosted on my computer. 

The likely only possible way to get this file is by digging it out of the 3DS somehow (I'm not that experienced with the 3DS's internals to know if/how is that possible), since nearly all packet dumps will just have a 304 from the original BOSS server.

silver-volt4 committed:
feat: adopt dani's notification PR

silver-volt4 committed:
chore: update to match the newly-merged notifications PR

silver-volt4 committed:
Merge pull request #1 from SwapdoodleRevival/feat/impl-notifications-pr

Use common notification implementation

silver-volt4 committed:
feat: add grpc endpoint

Intended for the processing of Note IDs, which are hidden under a button in the app.
They use the same format as Game Builder Garage (https://github.com/kinnay/NintendoClients/wiki/Data-Store-Codes#game-builder-garage). Many thanks to DaniElectra for figuring that out.

silver-volt4 committed:
Merge pull request #2 from SwapdoodleRevival/feat/grpc

Add gRPC endpoint to retrieve Notes by Note IDs

jonbarrow commented:
It's a legitimate bot we use. It was just seemingly broken for a while and no one noticed, it was only fixed yesterday

The CLA is just an acknowledgment that you own the rights to propose the changes in the PR. We started using it after a few people tried to PR in information from Nintendo SDK leaks, as an effort to officially take a stance against that practice

After the bot got fixed it sent out a ton of messages from the backlog, so it looks a little spammy
